File: src/main/java/no/nav/pto/veilarbfilter/service/MetricsReporter.java
package no.nav.pto.veilarbfilter.service;

import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.MultiGauge;
import io.micrometer.core.instrument.Tags;
import io.micrometer.core.instrument.binder.MeterBinder;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.pto.veilarbfilter.domene.PortefoljeFilter;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class MetricsReporter implements MeterBinder {
    private final MineLagredeFilterService mineLagredeFilterService;

    private MultiGauge lagredeFilterStats;

    @Override
    public void bindTo(@NotNull MeterRegistry meterRegistry) {
        if (lagredeFilterStats == null) {
            lagredeFilterStats = MultiGauge.builder("portefolje_lagredefilter")
                    .description("Stats for lagrede filter")
                    .register(meterRegistry);
        }
    }

    public void reportLagradeFilter() {
        try {
            Map<String, Integer> stats = new HashMap<>();

            log.info("Reporting metrics...");

            mineLagredeFilterService.hentAllLagredeFilter().forEach(filterModel -> {

                PortefoljeFilter filterValg = filterModel.getFilterValg();
                if (filterValg.getAktiviteter() != null) {
                    incrementFilterStats(stats, "aktiviteter");
                    if (!"NA".equals(filterValg.getAktiviteter().getBEHANDLING())) {
                        incrementFilterStats(stats, "behandling");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getEGEN())) {
                        incrementFilterStats(stats, "egen");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getGRUPPEAKTIVITET())) {
                        incrementFilterStats(stats, "gruppeaktivitet");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getIJOBB())) {
                        incrementFilterStats(stats, "ijobb");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getMOTE())) {
                        incrementFilterStats(stats, "mote");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getSOKEAVTALE())) {
                        incrementFilterStats(stats, "sokeavtale");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getSTILLING())) {
                        incrementFilterStats(stats, "stilling");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getTILTAK())) {
                        incrementFilterStats(stats, "tiltak");
                    }
                    if (!"NA".equals(filterValg.getAktiviteter().getUTDANNINGAKTIVITET())) {
                        incrementFilterStats(stats, "utdanningaktivitet");
                    }
                }
                if (filterValg.getAlder() != null && !filterValg.getAlder().isEmpty()) {
                    incrementFilterStats(stats, "alder");
                }
                if (filterValg.getFerdigfilterListe() != null && !filterValg.getFerdigfilterListe().isEmpty()) {
                    incrementFilterStats(stats, "ferdigfilterliste");
                    filterValg.getFerdigfilterListe().forEach(x -> incrementFilterStats(stats, x));
                }
                if (filterValg.getFodselsdagIMnd() != null && !filterValg.getFodselsdagIMnd().isEmpty()) {
                    incrementFilterStats(stats, "fodselsdagimnd");
                }
                if (filterValg.getFormidlingsgruppe() != null && !filterValg.getFormidlingsgruppe().isEmpty()) {
                    incrementFilterStats(stats, "formidlingsgruppe");
                }
                if (filterValg.getKjonn() != null && !filterValg.getKjonn().isEmpty()) {
                    incrementFilterStats(stats, "kjonn");
                }
                if (filterValg.getManuellBrukerStatus() != null && !filterValg.getManuellBrukerStatus().isEmpty()) {
                    incrementFilterStats(stats, "manuellbrukerstatus");
                }
                if (filterValg.getRettighetsgruppe() != null && !filterValg.getRettighetsgruppe().isEmpty()) {
                    incrementFilterStats(stats, "rettighetsgruppe");
                }
                if (filterValg.getServicegruppe() != null && !filterValg.getServicegruppe().isEmpty()) {
                    incrementFilterStats(stats, "servicegruppe");
                }
                if (filterValg.getTiltakstyper() != null && !filterValg.getTiltakstyper().isEmpty()) {
                    incrementFilterStats(stats, "tiltakstyper");
                }
                if (filterValg.getVeilederNavnQuery() != null && !filterValg.getVeilederNavnQuery().isEmpty()) {
                    incrementFilterStats(stats, "veiledernavnquery");
                }
                if (filterValg.getVeiledere() != null && !filterValg.getVeiledere().isEmpty()) {
                    incrementFilterStats(stats, "veiledere");
                }
                if (filterValg.getRegistreringstype() != null && !filterValg.getRegistreringstype().isEmpty()) {
                    incrementFilterStats(stats, "registreringstype");
                }
                if (filterValg.getCvJobbprofil() != null && !filterValg.getCvJobbprofil().isEmpty()) {
                    incrementFilterStats(stats, "cvjobbprofil");
                }
                if (filterValg.getLandgruppe() != null && !filterValg.getLandgruppe().isEmpty()) {
                    incrementFilterStats(stats, "landgruppe");
                }
                if (filterValg.getFoedeland() != null && !filterValg.getFoedeland().isEmpty()) {
                    incrementFilterStats(stats, "foedeland");
                }
                if (filterValg.getBarnUnder18Aar() != null && !filterValg.getBarnUnder18Aar().isEmpty()) {
                    incrementFilterStats(stats, "barnunder18aar");
                }
                if (filterValg.getTolkebehov() != null && !filterValg.getTolkebehov().isEmpty()) {
                    incrementFilterStats(stats, "tolkbehov");
                }
                if (filterValg.getEnsligeForsorgere() != null && !filterValg.getEnsligeForsorgere().isEmpty()) {
                    incrementFilterStats(stats, "ensligeforsorgere");
                }
                if (filterValg.getFargekategorier() != null && !filterValg.getFargekategorier().isEmpty()) {
                    incrementFilterStats(stats, "fargekategorier");
                }
            });
            lagredeFilterStats.register(stats.entrySet().stream().map(entry -> MultiGauge.Row.of(Tags.of("filterNavn", entry.getKey()), entry.getValue())).collect(Collectors.toList()), true);
        } catch (Exception e) {
            log.error("Can not report metrics " + e, e);
        }
    }

    private void incrementFilterStats(Map<String, Integer> stats, String statName) {
        if (stats.containsKey(statName)) {
            stats.put(statName, stats.get(statName) + 1);
        } else {
            stats.put(statName, 1);
        }
    }
}

File: src/main/no/nav/pto/veilarbfilter/service/MetricsReporter.kt
package no.nav.pto.veilarbfilter.service

import no.nav.common.metrics.Event
import no.nav.common.metrics.InfluxClient
import no.nav.common.metrics.MetricsClient
import no.nav.pto.veilarbfilter.config.dbQuery
import no.nav.pto.veilarbfilter.db.MetricsReporterInfo
import org.apache.commons.codec.digest.DigestUtils
import org.jetbrains.exposed.sql.select
import org.jetbrains.exposed.sql.update
import org.slf4j.LoggerFactory
import java.time.LocalDate
import java.time.LocalDateTime

class MetricsReporterService {
    private var mineLagredeFilterServiceImpl: MineLagredeFilterServiceImpl
    private var metricsClient: MetricsClient
    private val log = LoggerFactory.getLogger("MetricsReporterService")
    private val REPORT_ID: Int = 1;

    constructor(mineLagredeFilterServiceImpl: MineLagredeFilterServiceImpl) {
        this.mineLagredeFilterServiceImpl = mineLagredeFilterServiceImpl
        metricsClient = InfluxClient()
    }

    private suspend fun shouldReport(): Boolean {
        var lastReportedTime = lastReportedTime();
        if (lastReportedTime!!.toLocalDate().isEqual(LocalDate.now())) {
            return false;
        }
        return true
    }

    private suspend fun saveLastReportingTime() {
        dbQuery {
            MetricsReporterInfo
                .update({ MetricsReporterInfo.reporterId eq REPORT_ID }) {
                    it[opprettetDato] = LocalDateTime.now()
                }
        }
    }

    private suspend fun lastReportedTime(): LocalDateTime? = dbQuery {
        (MetricsReporterInfo).slice(
            MetricsReporterInfo.opprettetDato,
        ).select { (MetricsReporterInfo.reporterId eq REPORT_ID) }
            .mapNotNull { row -> row[MetricsReporterInfo.opprettetDato] }
            .singleOrNull()
    }

    suspend fun reportLagradeFilter() {

        if (!shouldReport()) {
            return;
        }

        saveLastReportingTime();

        log.info("Reporting metrics...")

        mineLagredeFilterServiceImpl.hentAllLagredeFilter().forEach {

            val metrikk = Event("portefolje.metrikker.lagredefilter.veileder-filter-counter")
            metrikk.addFieldToReport("navn-lengde", it.filterNavn.length)
            metrikk.addTagToReport("id", getHash(it.veilederId))
            metrikk.addFieldToReport("filterId", it.filterId)
            val filterValg = it.filterValg
            var antallFiltre = 0

            if (filterValg.aktiviteter != null) {
                metrikk.addTagToReport("aktiviteter", "1")
                if (filterValg.aktiviteter.BEHANDLING != "NA") {
                    metrikk.addTagToReport("BEHANDLING", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.EGEN != "NA") {
                    metrikk.addTagToReport("EGEN", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.GRUPPEAKTIVITET != "NA") {
                    metrikk.addTagToReport("GRUPPEAKTIVITET", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.IJOBB != "NA") {
                    metrikk.addTagToReport("IJOBB", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.MOTE != "NA") {
                    metrikk.addTagToReport("MOTE", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.SOKEAVTALE != "NA") {
                    metrikk.addTagToReport("SOKEAVTALE", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.STILLING != "NA") {
                    metrikk.addTagToReport("STILLING", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.TILTAK != "NA") {
                    metrikk.addTagToReport("TILTAK", "1")
                    antallFiltre++
                }
                if (filterValg.aktiviteter.UTDANNINGAKTIVITET != "NA") {
                    metrikk.addTagToReport("UTDANNINGAKTIVITET", "1")
                    antallFiltre++
                }
            }
            if (filterValg.alder.isNotEmpty()) {
                metrikk.addTagToReport("alder", "1")
                addValuesAsTags(metrikk, filterValg.alder)
                antallFiltre += filterValg.alder.size
            }
            if (filterValg.ferdigfilterListe.isNotEmpty()) {
                metrikk.addTagToReport("ferdigfilterListe", "1")
                addValuesAsTags(metrikk, filterValg.ferdigfilterListe)
                antallFiltre += filterValg.ferdigfilterListe.size
            }
            if (filterValg.fodselsdagIMnd.isNotEmpty()) {
                metrikk.addTagToReport("fodselsdagIMnd", "1")
                antallFiltre += filterValg.fodselsdagIMnd.size
            }
            if (filterValg.formidlingsgruppe.isNotEmpty()) {
                metrikk.addTagToReport("formidlingsgruppe", "1")
                addValuesAsTags(metrikk, filterValg.formidlingsgruppe)
                antallFiltre += filterValg.formidlingsgruppe.size
            }
            if (filterValg.hovedmal.isNotEmpty()) {
                metrikk.addTagToReport("hovedmal", "1")
                addValuesAsTags(metrikk, filterValg.hovedmal)
                antallFiltre += filterValg.hovedmal.size
            }
            if (filterValg.innsatsgruppe.isNotEmpty()) {
                metrikk.addTagToReport("innsatsgruppe", "1")
                addValuesAsTags(metrikk, filterValg.innsatsgruppe)
                antallFiltre += filterValg.innsatsgruppe.size
            }
            if (filterValg.kjonn != null && filterValg.kjonn.isNotEmpty()) {
                metrikk.addTagToReport("kjonn", "1")
                metrikk.addTagToReport(filterValg.kjonn, "1")
                antallFiltre++
            }
            if (filterValg.manuellBrukerStatus.isNotEmpty()) {
                metrikk.addTagToReport("manuellBrukerStatus", "1")
                addValuesAsTags(metrikk, filterValg.manuellBrukerStatus)
                antallFiltre += filterValg.manuellBrukerStatus.size
            }
            if (filterValg.rettighetsgruppe.isNotEmpty()) {
                metrikk.addTagToReport("rettighetsgruppe", "1")
                addValuesAsTags(metrikk, filterValg.rettighetsgruppe)
                antallFiltre += filterValg.rettighetsgruppe.size
            }
            if (filterValg.servicegruppe.isNotEmpty()) {
                metrikk.addTagToReport("servicegruppe", "1")
                addValuesAsTags(metrikk, filterValg.servicegruppe)
                antallFiltre += filterValg.servicegruppe.size
            }
            if (filterValg.tiltakstyper.isNotEmpty()) {
                metrikk.addTagToReport("tiltakstyper", "1")
                addValuesAsTags(metrikk, filterValg.tiltakstyper)
                antallFiltre += filterValg.tiltakstyper.size
            }
            if (filterValg.veilederNavnQuery.isNotEmpty()) {
                metrikk.addTagToReport("veilederNavnQuery", "1")
                antallFiltre++
            }
            if (filterValg.veiledere.isNotEmpty()) {
                metrikk.addTagToReport("veiledere", "1")
                antallFiltre += filterValg.veiledere.size
            }
            if (filterValg.ytelse != null && filterValg.ytelse.isNotEmpty()) {
                metrikk.addTagToReport("ytelse", "1")
                metrikk.addTagToReport(filterValg.ytelse, "1")
                antallFiltre++
            }
            if (filterValg.registreringstype != null && filterValg.registreringstype.isNotEmpty()) {
                metrikk.addTagToReport("registreringstype", "1")
                addValuesAsTags(metrikk, filterValg.registreringstype)
                antallFiltre += filterValg.registreringstype.size
            }
            if (filterValg.cvJobbprofil != null && filterValg.cvJobbprofil.isNotEmpty()) {
                metrikk.addTagToReport("cvJobbprofil", "1")
                metrikk.addTagToReport(filterValg.cvJobbprofil, "1")
                antallFiltre++
            }
            if (filterValg.arbeidslisteKategori != null && filterValg.arbeidslisteKategori.isNotEmpty()) {
                metrikk.addTagToReport("arbeidslisteKategori", "1")
                antallFiltre += filterValg.arbeidslisteKategori.size
            }
            metrikk.addFieldToReport("antallFiltre", antallFiltre)
            metricsClient.report(metrikk)
        }
    }

    private fun addValuesAsTags(metrikk: Event, listValues: List<String>) {
        listValues.forEach { metrikk.addTagToReport(it, "1") }
    }

    private fun getHash(veilederId: String): String {
        return DigestUtils.md5Hex(veilederId)
    };
}